### DISCUSSION TITLE
Infeasible_Problem_Detected

### DISCUSSION THREAD

**Discussion by TheSimplyRed** [coin-or/Ipopt#789] [Q&A]
Hi, 
How Can I find why Ipopt considers a solution as infeasible, when I see that it is feasible considering g_L, and g_U?
My g_L, g_U, and g (with "infeasible" x)
[Feasible result.xlsx](https://github.com/user-attachments/files/16910455/Feasible.result.xlsx)
A detailed output from ipopt
[detailed.txt](https://github.com/user-attachments/files/16910445/detailed.txt)

**Comment by svigerske** [comment]
There seems to be a lot of lines missing in `detailed.txt`. It also contains
```
*** Output truncated by debugger ***
```
For the last iteration that is shown in the log, Ipopt is in restoration phase at a point with infeasibility 1.23e-04 in the original NLP. The restoration phase NLP is solved to local optimality at this point, so there is no way to reduce the infeasibility in the original problem. Maybe the point you get returned from Ipopt is from before the restoration phase was entered. But since the log is so incomplete, I cannot see at which point Ipopt was at this point and why it decided to enter restoration phase.

**Reply by TheSimplyRed** [reply]
The output was in VS, with output level 10 the size of the file is huge.
Here are the last 10 iterations 
[Last10.txt](https://github.com/user-attachments/files/16930222/Last10.txt)
Here is the file with all iterations and level 6 
[ipopt6.txt](https://github.com/user-attachments/files/16930235/ipopt6.txt)

**Reply by svigerske** [reply]
Thanks. Looks like the point where it went into the restoration phase was indeed infeasible and the one that is returned to you at the end is the one from the end of the restoration phase.
That is, the trial point with the 2.17e-06 from the line
```
orig_curr_inf_pr = 1.23e-04, orig_trial_inf_pr = 2.17e-06
```

Now, just some explanation why there are different values for the primal infeasibility.
If you think of the inequality and equality constraints that are passed to Ipopt as
```
d_L <= d(x) <= d_U
c(x) = 0
```
then Ipopt internally solves this as
```
d(x) - s = 0
c(x) = 0
d_L <= s <= d_U
```
The infeasibility 1.23e-04 is a violation of d(x)-s:
```
curr_d_minus_s[   18]= 2.1666492326077013e-06
```
The infeasibility of 3.1e-9 from the output
```
Constraint violation....:   3.1160151907846598e-09    3.1160151907846598e-09
```
is measured by evaluating ||d(x)-d_L||_inf, ||d_U-d(x)||_inf.

So it seems that the value of s is not right yet, which means that the solution for the internal NLP is not feasible yet. Because the restoration phase converged, Ipopt thought it is in a local infeasible point and then claimed that also the original NLP is infeasible.

**Reply by TheSimplyRed** [reply]
Thanks for your answer.
We are trying different ways to resolve our problem.
All our constraints are supposed to be equality constraints, we use +- epsilon for some of them during our tests.
In the example that I sent, what I am supposed to do to make Ipopt consider the point where it stopped as a good solution?
Is it related to tolerance?
I made a new calculation with the option AddIpoptNumOption(nlp, (char*)"tol", 0.01); 
I got STEP_COMPUTATION_FAILED
[ipopt12.txt](https://github.com/user-attachments/files/16947440/ipopt12.txt)

**Reply by svigerske** [reply]
Then I suspect that these +-epsilon are the problem. An equality will be handled a such, but these inequalities create slack variables with bounds [-epsilon,+epsilon]. Ipopt is an **interior point** algorithm, so this gives a very tiny interior.

You could introduce the slack variable explicitly on your side.
Also check whether you really need to do these tests. Maybe letting the slacks free but adding a term to the objective that minimizes the sum of the squares of the slacks will do better?

**Reply by TheSimplyRed** [reply]
We modified all inequality constraints with slack variables, so that all our constraints became  equality constraints.
In that example 
[ipopt8.txt](https://github.com/user-attachments/files/17021164/ipopt8.txt)
 we got restoration  phase failed after the first iteration.
I tried to used output level 12 then 10, it failed with memory access violation.

**Reply by TheSimplyRed** [reply]
> We modified all inequality constraints with slack variables, so that all our constraints became equality constraints. In that example [ipopt8.txt](https://github.com/user-attachments/files/17021164/ipopt8.txt) we got restoration phase failed after the first iteration. I tried to used output level 12 then 10, it failed with memory access violation.

The issue with restoration phase failed is resolved, We forgot to initialize the new slack variables.

**Comment by TheSimplyRed** [comment]
We did a calculation with random initial values, it converged to an acceptable solution.
We took the solution and used it as initial value, we used warm start option, we were expecting to have a fast convergences (few iterations), but it took more time to converge than with the random initial values.
the link for the first log https://www.dropbox.com/scl/fi/vm46f4uzmgqrh0euete6m/ipopt8_NoWarmStart.txt?rlkey=dbnqpqpe2mczglabqqlxynwce&dl=0
the link for the log with warm start https://www.dropbox.com/scl/fi/rbe5k65fho4oos8d0w3jc/ipopt8_warm-Start.txt?rlkey=jj417tegpvuv5g8720mgodht8&dl=0
